docs/conf.py

```diff
@@ -294,7 +294,11 @@
 # -- Extension configuration -------------------------------------------------
 
 # Tell sphinx autodoc how to render type aliases.
-autodoc_typehints = "description"
+
+# type aliases are not respected in overloaded function signatures, which leads to
```

Comment:
Is there an autodoc bug we can link to here?

Reply:
No, i don’t think it’s a bug. The type signatures are displaying correctly, they’re just not helpful in most cases because of our long aliases.

Reply:
I think autodoc does not expand aliases in modules with `from __future__ import annotations`. 

We did this hack in Haiku because of that: https://github.com/google-deepmind/dm-haiku/blob/a7b7e73dae840153ecd828e97a64b6a875b168f7/docs/conf.py#L36, but since JAX is 3.9+ we can also just add the `__future__` import everywhere. Wdyt?

Reply:
I saw that in the docs. We already have the future import in the relevant modules, it didn't seem to make a difference.

Reply:
The aliasing works correctly in the `description`, but seems not to work in overloads.

We previously had `autodoc_typehints='description'`, but according to the [documentation](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_typehints), for overloaded functions this is the same as `autodoc_typehints='signature'`.

I think our only recourse to avoid the busy expansions in the `jnp.where` screenshot above is to set `autodoc_typehints='none'`.

Reply:
I've sent #22208 which seems to work locally.